File: csharp/demo/demo20a03-semaphore-deadlock.cs
/*
 * SEMAPHORES
 * Version A: Paper sheets and packages
 */
using System;
using System.Threading;



class Demo20A03 : IRunnable
{
    public void run()
    {
        var semPackage = new Semaphore(0, int.MaxValue);
        var semSheet = new Semaphore(2, int.MaxValue);


        ThreadStart makeOneSheet = () =>
        {
            for (int i = 0; i < 4; ++i)
            {
                semSheet.WaitOne();
                Console.WriteLine("Make 1 sheet");
                semPackage.Release();
            }
        };


        ThreadStart combineOnePackage = () =>
        {
            for (int i = 0; i < 4; ++i)
            {
                semPackage.WaitOne();
                semPackage.WaitOne();

                Console.WriteLine("Combine 2 sheets into 1 package");
                Thread.Sleep(1000);

                semSheet.Release();
                // Missing one statement: semSheet.Release() ==> deadlock
            }
        };


        new Thread(makeOneSheet).Start();
        new Thread(makeOneSheet).Start();
        new Thread(combineOnePackage).Start();
    }
}

File: python/demo20a03_semaphore_deadlock.py
'''
SEMAPHORES
Version A: Paper sheets and packages
'''

import time
import threading



sem_package = threading.Semaphore(0)
sem_sheet = threading.Semaphore(2)



def make_one_sheet():
    for _ in range(4):
        sem_sheet.acquire()
        print('Make 1 sheet')
        sem_package.release()



def combine_one_package():
    for _ in range(4):
        sem_package.acquire()
        sem_package.acquire()
        print('Combine 2 sheets into 1 package')
        time.sleep(2)
        sem_sheet.release()
        # sem_sheet.release() # Missing one statement: sem_sheet.release() ==> deadlock



threading.Thread(target=make_one_sheet).start()
threading.Thread(target=make_one_sheet).start()
threading.Thread(target=combine_one_package).start()
